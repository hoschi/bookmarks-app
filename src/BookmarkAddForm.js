import { Grid, Button, TextField, withStyles, Paper } from '@material-ui/core'
import AddIcon from '@material-ui/icons/Add'
import { Form, withFormik } from 'formik'
import PropTypes from 'prop-types'
import * as R from 'ramda'
import React from 'react'
import { setPropTypes } from 'recompose'
import { getNewBookmarkId } from './gql'

export const getTextFieldDefaultProps = (
    id,
    { handleChange, handleBlur, values, errors, touched }
) => ({
    id,
    margin: 'normal',
    onChange: handleChange,
    onBlur: handleBlur,
    value: R.propOr('', id, values),
    error: !!(touched[id] && errors[id]),
    helperText: touched[id] && errors[id] ? errors[id] : undefined,
    fullWidth: true,
})

const styles = (theme) => ({
    root: {
        marginTop: 16,
        marginBottom: 80,
        position: 'relative',
        padding: '0 16px 16px 16px',
    },
    fields: {
        flex: 1,
    },
    buttonContainer: {
        width: 80,
        paddingLeft: 16,
        [theme.breakpoints.up('sm')]: {
            width: 88,
            paddingLeft: 24,
        },
    },
    button: {
        position: 'absolute',
        left: '50%',
        transform: 'translateX(-50%)',
        bottom: -68,
    },
})

function BookmarkAddForm(props) {
    const { classes, isValid, isSubmitting } = props
    return (
        <Paper className={classes.root}>
            <Form>
                <Grid container>
                    <Grid item className={classes.fields}>
                        <TextField {...getTextFieldDefaultProps('url', props)} label="URL" />
                        <TextField {...getTextFieldDefaultProps('title', props)} label="Title" />
                    </Grid>
                </Grid>
                <Button
                    color="primary"
                    variant="fab"
                    type="submit"
                    aria-label="Add"
                    className={classes.button}
                    disabled={!isValid || isSubmitting}
                >
                    <AddIcon />
                </Button>
            </Form>
        </Paper>
    )
}

BookmarkAddForm.propTypes = {
    classes: PropTypes.object.isRequired,
}

export const BookmarkAddFormStyled = withStyles(styles)(BookmarkAddForm)

export function validate(values) {
    let errors = {}
    if (!values.title) {
        errors.title = 'Required'
    }
    if (!values.url) {
        errors.url = 'Required'
    }
    return errors
}

export function handleSubmit(values, { setSubmitting, resetForm, setErrors, props }) {
    setSubmitting(true)
    const { createBookmark } = props
    return createBookmark({
        variables: {
            ...values,
            id: getNewBookmarkId(),
        },
    })
        .then(() => {
            resetForm()
            setSubmitting(false)
        })
        .catch((err) => {
            const valuesKeys = R.keys(values)
            setErrors(R.zipObj(valuesKeys, R.times(R.always(err.message), valuesKeys.length)))
            setSubmitting(false)
        })
}

export default R.pipe(
    withFormik({
        validate,
        handleSubmit,
    }),
    setPropTypes({
        createBookmark: PropTypes.func.isRequired,
    })
)(BookmarkAddFormStyled)
