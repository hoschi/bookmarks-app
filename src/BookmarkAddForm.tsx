import AddIcon from '@mui/icons-material/Add'
import { Paper, Fab, Stack } from '@mui/material'
import { ComponentProps, useState } from 'react'
import * as R from 'ramda'
import { createBookmark as createBookmarkMutation, queryAllBookmarks } from './gql'
import { TextFieldElement as TextFieldElementBase, useForm } from 'react-hook-form-mui'
import { useMutation } from '@apollo/client'

interface FormValues {
    url: string
    title: string
}
const TextFieldElement = TextFieldElementBase<FormValues>

const TEXT_FIELD_PROPS: Partial<ComponentProps<typeof TextFieldElement>> = {
    required: true,
    fullWidth: true,
    margin: 'normal',
} as const

export const BookmarkAddForm: React.FC = () => {
    const [isSubmitting, setSubmitting] = useState<boolean>(false)
    const [createBookmark, { error: mutationError }] = useMutation(createBookmarkMutation, {
        refetchQueries: [queryAllBookmarks],
    })

    if (mutationError) throw new Error(mutationError.message)
    const { control, handleSubmit, reset, setError } = useForm<FormValues>({
        defaultValues: {
            url: '',
            title: '',
        },
    })

    return (
        <Paper sx={{ mt: 2, mb: 10, p: 2, pt: 0, position: 'relative' }}>
            <form
                noValidate
                onSubmit={handleSubmit(async (values: FormValues) => {
                    setSubmitting(true)
                    try {
                        await createBookmark({
                            variables: values,
                        })

                        reset()
                        setSubmitting(false)
                    } catch (e: unknown) {
                        const err: object = e || {}
                        const errorMessage: string =
                            'message' in err && err.message instanceof String
                                ? err.message.toString()
                                : 'Unknown error'
                        R.forEach(
                            (key: keyof FormValues) =>
                                setError(key, { type: 'validate', message: errorMessage }),
                            R.keys(values)
                        )
                        setSubmitting(false)
                    }
                })}
            >
                <Stack>
                    <TextFieldElement
                        {...TEXT_FIELD_PROPS}
                        control={control}
                        name="url"
                        label="URL"
                    />
                    <TextFieldElement
                        {...TEXT_FIELD_PROPS}
                        control={control}
                        name="title"
                        label="Title"
                    />
                </Stack>
                <Fab
                    color="primary"
                    type="submit"
                    aria-label="Add"
                    disabled={isSubmitting}
                    sx={{
                        position: 'absolute',
                        left: '50%',
                        transform: 'translateX(-50%)',
                        bottom: -68,
                    }}
                >
                    <AddIcon />
                </Fab>
            </form>
        </Paper>
    )
}
