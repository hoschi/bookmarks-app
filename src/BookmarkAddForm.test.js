import * as R from 'ramda'
import React from 'react'
import { shallow } from 'enzyme'
import { BookmarkAddFormStyled, validate, handleSubmit } from './BookmarkAddForm'

function render(props) {
    return shallow(<BookmarkAddFormStyled {...props} />).dive()
}

test('valid, submitting', function() {
    const actual = render({
        isValid: true,
        isSubmitting: true,
        values: { title: 'my title', url: 'http://foo.com' },
        touched: { title: true, url: true },
        errors: {},
    })
    expect(actual).toMatchSnapshot()
})

test('valid, not submitting', function() {
    const actual = render({
        isValid: true,
        isSubmitting: false,
        values: { title: 'my title', url: 'http://foo.com' },
        touched: { title: true, url: true },
        errors: {},
    })
    expect(actual).toMatchSnapshot()
})

test('not valid, not submitting', function() {
    const actual = render({
        isValid: false,
        isSubmitting: true,
        values: { title: '', url: '' },
        touched: { title: true, url: true },
        errors: {},
    })
    expect(actual).toMatchSnapshot()
})

test('initial errors', function() {
    const actual = render({
        isValid: false,
        isSubmitting: false,
        values: { title: '', url: '' },
        touched: { title: false, url: false },
        errors: { title: 'Required!', url: 'Required!' },
    })
    expect(actual).toMatchSnapshot()
})

test('show errors', function() {
    const actual = render({
        isValid: false,
        isSubmitting: false,
        values: { title: '', url: '' },
        touched: { title: true, url: true },
        errors: { title: 'Required!', url: 'Required!' },
    })
    expect(actual).toMatchSnapshot()
})

test('validate', function() {
    expect(validate({ title: '', url: '' })).toEqual({
        title: 'Required',
        url: 'Required',
    })
    expect(validate({ title: 'my title' })).toEqual({
        url: 'Required',
    })
    expect(validate({ url: 'http://foo-bar.com' })).toEqual({
        title: 'Required',
    })
    expect(validate({ title: 'my title', url: 'http://foo-bar.com' })).toEqual({})
})

describe('handleSubmit', function() {
    const valuesIn = {
        title: 'my title',
        url: 'https://you.rock',
    }
    const badge = {
        setSubmitting: jest.fn(),
        resetForm: jest.fn(),
        setErrors: jest.fn(),
    }

    beforeEach(function() {
        badge.setSubmitting.mockClear()
        badge.resetForm.mockClear()
        badge.setErrors.mockClear()
    })

    it('happy case', function() {
        expect.assertions(5)
        return handleSubmit(valuesIn, {
            ...badge,
            props: {
                createBookmark: (mutationParams) => {
                    expect(badge.setSubmitting).toHaveBeenLastCalledWith(true)
                    expect(R.omit(['id'], mutationParams.variables)).toEqual(valuesIn)
                    return new Promise((resolve) => {
                        resolve()
                    })
                },
            },
        }).then(() => {
            expect(badge.resetForm).toHaveBeenCalled()
            expect(badge.setErrors).not.toHaveBeenCalled()
            expect(badge.setSubmitting).toHaveBeenLastCalledWith(false)
        })
    })

    it('error from server', function() {
        expect.assertions(5)
        const msg = 'Uh oh!'
        return handleSubmit(valuesIn, {
            ...badge,
            props: {
                createBookmark: (mutationParams) => {
                    expect(badge.setSubmitting).toHaveBeenLastCalledWith(true)
                    expect(R.omit(['id'], mutationParams.variables)).toEqual(valuesIn)
                    return new Promise((resolve, reject) => {
                        reject(new Error(msg))
                    })
                },
            },
        }).then(() => {
            expect(badge.resetForm).not.toHaveBeenCalled()
            expect(badge.setErrors).toHaveBeenLastCalledWith({
                title: msg,
                url: msg,
            })
            expect(badge.setSubmitting).toHaveBeenLastCalledWith(false)
        })
    })
})
