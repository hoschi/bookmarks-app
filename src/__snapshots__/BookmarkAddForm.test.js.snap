// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial errors 1`] = `
<WithStyles(Paper)
  className="BookmarkAddForm-root-1"
>
  <Form>
    <WithStyles(Grid)
      container={true}
    >
      <WithStyles(Grid)
        className="BookmarkAddForm-fields-2"
        item={true}
      >
        <TextField
          error={false}
          fullWidth={true}
          id="url"
          label="URL"
          margin="normal"
          required={false}
          select={false}
          value=""
          variant="standard"
        />
        <TextField
          error={false}
          fullWidth={true}
          id="title"
          label="Title"
          margin="normal"
          required={false}
          select={false}
          value=""
          variant="standard"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Button)
      aria-label="Add"
      className="BookmarkAddForm-button-4"
      color="primary"
      disabled={true}
      type="submit"
      variant="fab"
    >
      <pure(AddIcon) />
    </WithStyles(Button)>
  </Form>
</WithStyles(Paper)>
`;

exports[`not valid, not submitting 1`] = `
<WithStyles(Paper)
  className="BookmarkAddForm-root-1"
>
  <Form>
    <WithStyles(Grid)
      container={true}
    >
      <WithStyles(Grid)
        className="BookmarkAddForm-fields-2"
        item={true}
      >
        <TextField
          error={false}
          fullWidth={true}
          id="url"
          label="URL"
          margin="normal"
          required={false}
          select={false}
          value=""
          variant="standard"
        />
        <TextField
          error={false}
          fullWidth={true}
          id="title"
          label="Title"
          margin="normal"
          required={false}
          select={false}
          value=""
          variant="standard"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Button)
      aria-label="Add"
      className="BookmarkAddForm-button-4"
      color="primary"
      disabled={true}
      type="submit"
      variant="fab"
    >
      <pure(AddIcon) />
    </WithStyles(Button)>
  </Form>
</WithStyles(Paper)>
`;

exports[`show errors 1`] = `
<WithStyles(Paper)
  className="BookmarkAddForm-root-1"
>
  <Form>
    <WithStyles(Grid)
      container={true}
    >
      <WithStyles(Grid)
        className="BookmarkAddForm-fields-2"
        item={true}
      >
        <TextField
          error={true}
          fullWidth={true}
          helperText="Required!"
          id="url"
          label="URL"
          margin="normal"
          required={false}
          select={false}
          value=""
          variant="standard"
        />
        <TextField
          error={true}
          fullWidth={true}
          helperText="Required!"
          id="title"
          label="Title"
          margin="normal"
          required={false}
          select={false}
          value=""
          variant="standard"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Button)
      aria-label="Add"
      className="BookmarkAddForm-button-4"
      color="primary"
      disabled={true}
      type="submit"
      variant="fab"
    >
      <pure(AddIcon) />
    </WithStyles(Button)>
  </Form>
</WithStyles(Paper)>
`;

exports[`valid, not submitting 1`] = `
<WithStyles(Paper)
  className="BookmarkAddForm-root-1"
>
  <Form>
    <WithStyles(Grid)
      container={true}
    >
      <WithStyles(Grid)
        className="BookmarkAddForm-fields-2"
        item={true}
      >
        <TextField
          error={false}
          fullWidth={true}
          id="url"
          label="URL"
          margin="normal"
          required={false}
          select={false}
          value="http://foo.com"
          variant="standard"
        />
        <TextField
          error={false}
          fullWidth={true}
          id="title"
          label="Title"
          margin="normal"
          required={false}
          select={false}
          value="my title"
          variant="standard"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Button)
      aria-label="Add"
      className="BookmarkAddForm-button-4"
      color="primary"
      disabled={false}
      type="submit"
      variant="fab"
    >
      <pure(AddIcon) />
    </WithStyles(Button)>
  </Form>
</WithStyles(Paper)>
`;

exports[`valid, submitting 1`] = `
<WithStyles(Paper)
  className="BookmarkAddForm-root-1"
>
  <Form>
    <WithStyles(Grid)
      container={true}
    >
      <WithStyles(Grid)
        className="BookmarkAddForm-fields-2"
        item={true}
      >
        <TextField
          error={false}
          fullWidth={true}
          id="url"
          label="URL"
          margin="normal"
          required={false}
          select={false}
          value="http://foo.com"
          variant="standard"
        />
        <TextField
          error={false}
          fullWidth={true}
          id="title"
          label="Title"
          margin="normal"
          required={false}
          select={false}
          value="my title"
          variant="standard"
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
    <WithStyles(Button)
      aria-label="Add"
      className="BookmarkAddForm-button-4"
      color="primary"
      disabled={true}
      type="submit"
      variant="fab"
    >
      <pure(AddIcon) />
    </WithStyles(Button)>
  </Form>
</WithStyles(Paper)>
`;
